demo.adoc

```diff
@@ -0,0 +1,253 @@
+
+== Build a Spring Boot Docker Image
+
+. Build your app and send the image to your local Docker daemon:
++
+[source,shell]
+----
+./gradlew bootBuildImage --imageName=springbootdemo
+----
++
+NOTE: The `--imageName` parameter allows specifying an image name. Without it, the name would be something like `demospringboot:0.0.1-SNAPSHOT`.
+
+. Start your containerized web application with Docker:
++
+[source,shell]
+----
+docker run -it -p8080:8080 springbootdemo
+----
+
+. Make sure it works at `http://localhost:8080`.
+
+== Secure Your Spring Boot Application in Docker
+
+User management is never an easy task and, most certainly, is not the main objective of your application. Okta is an identity provider that helps you to take care of routine work such as:
+
+- Implementing OAuth 2.0
+- Adding Social Login
+- Implementing SSO (Single Sign-On)
+
+Okta is developer-friendly, and it has excellent integration with different frameworks, including Spring Boot.
+
+. Install the https://cli.okta.com[Okta CLI].
+. Run `okta register` if you don't have an account; `okta login` if you do.
+. Run `okta apps create web`.
+. Use the following redirect URIs:
+- Login: `http://localhost:8080/login/oauth2/code/okta`
+- Logout: `http://localhost:8080`
+. You'll need to run `source .okta.env` to set these values as environment variables.
+
+TIP: If you're on Windows, rename the file to `.okta.bat` and change `export` to `set`.
+
+=== Configure Spring Security to Lock Down Access
+
+. Modify your security settings in `DemoApplication` to allow access for authorized users only.
++
+[source,kotlin]
+----
+@Configuration
+class OktaOAuth2WebSecurityConfigurerAdapter: WebSecurityConfigurerAdapter() {
+    override fun configure(http: HttpSecurity) {
+        http.authorizeRequests().anyRequest().authenticated();
+    }
+}
+----
++
+That's it. The Okta Spring Boot Starter takes care of the rest!
+
+. Rebuild the application:
++
+[source,shell]
+----
+./gradlew bootBuildImage --imageName=springbootdemo
+----
+
+=== Start Spring Boot Application in Docker
+
+. Start your application in Docker with your Okta settings as environment variables:
++
+[source,shell]
+----
+docker run -it -p8080:8080 \
+  -e OKTA_OAUTH2_ISSUER="https://dev-xxxxxx.okta.com/oauth2/default" \
+  -e OKTA_OAUTH2_CLIENT_SECRET="yyyyyyyyyyyyyyyyyyy" \
+  -e OKTA_OAUTH2_CLIENT_ID="zzzzzzzzzzzzzzzz" \
+  springbootdemo
+----
++
+The argument `-e` allows to set an environment variable for the application running _inside_ your container and `-p` maps container's ports to `localhost` ports.
+
+. Navigate to `http://localhost:8080` in your browser and log in using your Okta credentials.
+
+. If you want to print the user's name (instead of the `sub` claim), use `@AuthenticatedPrincipal` and `OidcUser`:
++
+[source,kotlin]
+----
+fun home(@AuthenticationPrincipal user: OidcUser?) = "Welcome, ${user?.fullName ?: "guest"}!"
+----
+
+== Use a dotenv File for Environment Variables
```

Comment:
question: do you ever use code formatting in headers? `dotenv` ??

Reply:
Yes, I do. dotenv is a library name though, and I rarely put those in backticks. https://www.npmjs.com/package/dotenv